README.txt

CONTENTS OF THIS FILE
---------------------

 * About Drupal
 * Configuration and features
 * Installation profiles
 * Appearance
 * Developing for Drupal

ABOUT DRUPAL
------------
Drupal is an open source content management platform supporting a variety of
websites ranging from personal weblogs to large community-driven websites. For
more information, see the Drupal website at http://drupal.org/, and join the
Drupal community at http://drupal.org/community.

Legal information about Drupal:
 * Know your rights when using Drupal:
   See LICENSE.txt in the same directory as this document.
 * Learn about the Drupal trademark and logo policy:
   http://drupal.com/trademark

CONFIGURATION AND FEATURES
--------------------------

Drupal core (what you get when you download and extract a drupal-x.y.tar.gz or
drupal-x.y.zip file from http://drupal.org/project/drupal) has what you need to
get started with your website. It includes several modules (extensions that add
functionality) for common website features, such as managing content, user
accounts, image uploading, and search. Core comes with many options that allow
site-specific configuration. In addition to the core modules, there are
thousands of contributed modules (for functionality not included with Drupal
core) available for download.

More about configuration:
 * Install, upgrade, and maintain Drupal:
   See INSTALL.txt and UPGRADE.txt in the same directory as this document.
 * Learn about how to use Drupal to create your site:
   http://drupal.org/documentation
 * Download contributed modules to sites/all/modules to extend Drupal's
   functionality:
   http://drupal.org/project/modules
 * See also: "Developing for Drupal" for writing your own modules, below.

INSTALLATION PROFILES
---------------------

Installation profiles define additional steps (such as enabling modules,
defining content types, etc.) that run after the base installation provided
by core when Drupal is first installed. There are two basic installation
profiles provided with Drupal core.

Installation profiles from the Drupal community modify the installation process
to provide a website for a specific use case, such as a CMS for media
publishers, a web-based project tracking tool, or a full-fledged CRM for
non-profit organizations raising money and accepting donations. They can be
distributed as bare installation profiles or as "distributions". Distributions
include Drupal core, the installation profile, and all other required
extensions, such as contributed and custom modules, themes, and third-party
libraries. Bare installation profiles require you to download Drupal Core and
the required extensions separately; place the downloaded profile in the
/profiles directory before you start the installation process. Note that the
contents of this directory may be overwritten during updates of Drupal core;
it is advised to keep code backups or use a version control system.

Additionally, modules and themes may be placed inside subdirectories in a
specific installation profile such as profiles/your_site_profile/modules and
profiles/your_site_profile/themes respectively to restrict their usage to only
sites that were installed with that specific profile.

More about installation profiles and distributions:
 * Read about the difference between installation profiles and distributions:
   http://drupal.org/node/1089736
 * Download contributed installation profiles and distributions:
   http://drupal.org/project/distributions
 * Develop your own installation profile or distribution:
   http://drupal.org/developing/distributions

APPEARANCE
----------

In Drupal, the appearance of your site is set by the theme (themes are
extensions that set fonts, colors, and layout). Drupal core comes with several
themes. More themes are available for download, and you can also create your own
custom theme.

More about themes:
 * Download contributed themes to sites/all/themes to modify Drupal's
   appearance:
   http://drupal.org/project/themes
 * Develop your own theme:
   http://drupal.org/documentation/theme

DEVELOPING FOR DRUPAL
---------------------

Drupal contains an extensive API that allows you to add to and modify the
functionality of your site. The API consists of "hooks", which allow modules to
react to system events and customize Drupal's behavior, and functions that
standardize common operations such as database queries and form generation. The
flexible hook architecture means that you should never need to directly modify
the files that come with Drupal core to achieve the functionality you want;
instead, functionality modifications take the form of modules.

When you need new functionality for your Drupal site, search for existing
contributed modules. If you find a module that matches except for a bug or an
additional needed feature, change the module and contribute your improvements
back to the project in the form of a "patch". Create new custom modules only
when nothing existing comes close to what you need.

More about developing:
 * Search for existing contributed modules:
   http://drupal.org/project/modules
 * Contribute a patch:
   http://drupal.org/patch/submit
 * Develop your own module:
   http://drupal.org/developing/modules
 * Follow best practices:
   http://drupal.org/best-practices
 * Refer to the API documentation:
   http://api.drupal.org/api/drupal/7


CHANGELOG.txt

<<<<<<< HEAD
Drupal 7.43, 2016-02-24
-----------------------
- Fixed security issues (multiple vulnerabilities). See SA-CORE-2016-001.

Drupal 7.42, 2016-02-03
-----------------------
- Stopped invoking hook_flush_caches() on every cron run, since some modules
  use that hook for expensive operations that are only needed on cache clears.
- Changed the default .htaccess and web.config to block Composer-related files.
- Added static caching to module_load_include() to improve performance.
- Fixed double-encoding bugs in select field widgets provided by the Options
  module. The fix deprecates the 'strip_tags' property on option widgets and
  replaces it with a new 'strip_tags_and_unescape' property (minor data
  structure change).
- Improved MySQL 5.7 support by changing the MySQL database driver to stop
  using the ANSI SQL mode alias, which has different meanings for different
  MySQL versions.
- Fixed a regression introduced in Drupal 7.39 which prevented autocomplete
  functionality from working on servers that are not configured to
  automatically recognize index.php.
- Updated the Archive_Tar PEAR package to the latest 1.4.0 release, to fix bugs
  with tar file handling on various operating systems.
- Fixed fatal errors on node preview when a field is displayed in the node
  teaser but hidden in the full node view. The fix removes a
  field_attach_prepare_view() call from the node_preview() function since it is
  redundant with one in the node preview theme layer.
- Improved the description of the "Trimmed" format option on text fields
  (translatable string change, and minor UI and data structure change).
- Numerous small bug fixes.
- Numerous API documentation improvements.
- Additional automated test coverage.
=======

>>>>>>> 5933478d3eb4b18b27b640ba0da440cda3856baa

Drupal 7.41, 2015-10-21
-----------------------
- Fixed security issues (open redirect). See SA-CORE-2015-004.

Drupal 7.40, 2015-10-14
-----------------------
- Made Drupal's code for parsing .info files run much faster and use much less
  memory.
- Prevented drupal_http_request() from returning an error when it receives a
  201 through 206 HTTP status code.
- Added support for autoloading traits via the registry on sites running PHP
  5.4 or higher.
- Allowed the user-picture.tpl.php theme template to have HTML classes besides
  the default "user-picture" class printed in it (markup change).
- Fixed the URL text filter to convert e-mail addresses with plus signs into
  mailto: links.
- Added alternate text to file icons displayed by the File module, to improve
  accessibility (string change, and minor API addition to theme_file_icon()).
- Changed one-time login link failure messages to be displayed as errors or
  warnings as appropriate, rather than as regular status messages (minor UI
  change and data structure change).
- Changed the default settings.php configuration to exclude private files from
  the "404_fast_paths" behavior.
- Changed the page that displays filter tips for a particular text format, for
  example filter/tips/full_html, to return "page not found" or "access denied"
  if the format does not exist or the user does not have access to it. This
  change adds a new menu item to the Filter module's hook_menu() entry (minor
  data structure change).
- Added a new hook, hook_block_cid_parts_alter(), to allow modules to alter the
  cache keys used for caching a particular block.
- Made drupal_set_message() display and return messages when "0" is passed in
  as the message to set.
- Fixed non-functional "Files displayed by default" setting on file fields.
- The "worker callback" provided in hook_cron_queue_info() and the "finished"
  callback specified during batch processing can now be any PHP callable
  instead of just functions.
- Prevented drupal_set_time_limit() from decreasing the time limit in the case
  where the PHP maximum execution time is already unlimited.
- Changed the default thousand marker for numeric fields from a space ("1 000")
  to nothing ("1000") (minor UI change: https://www.drupal.org/node/1388376).
- Prevented malformed theme .info files (without a "name" key) from causing
  exceptions during menu rebuilds. If an .info file without a "name" key is
  found in a module or theme directory, Drupal will now use the module or
  theme's machine name as the display name instead.
- Made the format column in the {date_format_locale} database table
  case-sensitive, to match the equivalent column in the {date_formats} table.
- Fixed a bug in the Statistics module that caused JavaScript files attached to
  a node while it is being viewed to be omitted from the page.
- Added an optional 'project:' prefix that can be added to dependencies in a
  module's .info file to indicate which project the dependency resides in (API
  addition: https://www.drupal.org/node/2299747).
- Fixed various bugs that occurred after hooks were invoked early in the Drupal
  bootstrap and that caused module_implements() and drupal_alter() to cache an
  incomplete set of hook implementations for later use.
- Set the X-Content-Type-Options header to "nosniff" when possible, to prevent
  certain web browsers from picking an unsafe MIME type.
- Prevented the database API from executing multiple queries at once on MySQL,
  if the site's PHP version is new enough to do so. This is a secondary defense
  against SQL injection (API change: https://www.drupal.org/node/2463973).
- Fixed a bug in the Drupal 6 to Drupal 7 upgrade path which caused the upgrade
  to fail when there were multiple file records pointing to the same file.
- Numerous small bug fixes.
- Numerous API documentation improvements.
- Additional automated test coverage.

Drupal 7.39, 2015-08-19
-----------------------
- Fixed security issues (multiple vulnerabilities). See SA-CORE-2015-003.


Drupal 7.38, 2015-06-17
-----------------------
- Fixed security issues (multiple vulnerabilities). See SA-CORE-2015-002.

Drupal 7.37, 2015-05-07

platform/index.php
<?php

/**
 * This file only returns JSON for the API calls. The actual app + HTML is in index.html.
 */

setcookie('NO_CACHE', '1');

if(strtolower($_SERVER['HTTP_X_REQUESTED_WITH']) != 'xmlhttprequest'){
  header("Location: index.html#/diagram/live");
  die();
}

/**
 * Showcase of Pantheon's abilities.
 */
header('Content-type: application/json');
header("Cache-Control: no-cache, must-revalidate"); // HTTP/1.1
header("Expires: Sat, 26 Jul 1997 05:00:00 GMT"); // Date in the past

$curl_result = pantheon_curl('https://api.live.getpantheon.com/sites/self/bindings', NULL, 8443);
$attr_result = pantheon_curl('https://api.live.getpantheon.com/sites/self', NULL, 8443);
//print $attr_result['body'];
//var_dump($curl_result);
$json_array = json_decode($curl_result['body'], true);
//print json_encode($json_array, JSON_PRETTY_PRINT);
//die;

$output = new stdClass();
$output->servers = array();

// Hard code edge servers.
$edge = new stdClass();
$edge->id = 'AAAAAAAA';
$edge->type = 'edgeserver';
$edge->endpoint = 'edge1';
$edge->slave_of = NULL;
$output->servers[] = $edge;

if (!isset($_GET['env'])){
  $_GET['env'] = PANTHEON_ENVIRONMENT;
}

// Make a smaller JSON structure to return to the frontend.
foreach ($json_array as $key => $val) {


  if ($val['environment'] == $_GET['env'] && !($val['type'] == 'pingdom')) {
    if(!isset($val['failover'])){
      $type = $val['type'];
      if(isset($val['slave_of']) && $type == "dbserver"){
        $type = "slavedbserver";
      }
      $binding = new stdClass();
      $binding->id = (string)substr($key, 0, 8);
      $binding->endpoint = isset($val['endpoint']) ? $val['endpoint'] : NULL;
      $binding->type = $type;
      $binding->slave_of = isset($val['slave_of']) ? $val['slave_of'] : NULL;
      $output->servers[] = $binding;
    }
  }
}

print json_encode($output);


platform/js/app.js
App = Ember.Application.create();
App.env = "";
var maxWidth = 1200;
//Directory above webroot, please use trailing slash
var install_dir = "platform/";

var _marketing = {
    'edgeserver'   :["Edge Routing and Cache", "The exterior of Pantheon — the part that directly touches the public internet. The Edge has a built-in, ultra-fast cache that's automatically enabled for every site. It improves page load times for our customers and helps sites to cruise through viral traffic spikes without breaking a sweat."],
    'appserver'    :["Application Container", "The essence of a runtime container is a highly tuned PHP-FPM worker and its connections to the outside world. Incoming requests come via an nginx web server which handles requests for static assets, and passes dynamic requests to PHP."],
    'dbserver'     :["Database Server","The Database Service uses MariaDB and a container architecture similar to the Runtime Matrix to provision DBs and perform workflow operations. Instead of scaling via load-balancing, the DB layer can provide redundancy and horizontal scalability by supporting a self-healing replication topology, which is managed automatically."],
    'slavedbserver':["Failover Database Server(Slave)","The Database Service uses MariaDB and a container architecture similar to the Runtime Matrix to provision DBs and perform workflow operations. Instead of scaling via load-balancing, the DB layer can provide redundancy and horizontal scalability by supporting a self-healing replication topology, which is managed automatically."],
    'cacheserver'  :["Redis","Available for your applications to use in order to speed up processing."],
    'fileserver'   :["Pantheon File System","Our PFS (Pantheon File System) is a breakthrough in network-attached storage. It is backed by a self-healing elastic cluster architecture, and its advanced FUSE client rivals local on-disk filesystems for performance, thanks to a thoroughly modern leveldb caching layer."],
    'newrelic'     :["New Relic APM","It’s about gaining actionable, real-time business insights from the billions of metrics your software is producing, including user click streams, mobile activity, end user experiences and transactions.<img src='img/newrelic-graph.png'>"],
    'indexserver'  :["Apache Solr™","The popular, blazing fast open source enterprise search platform from the Apache Lucene project. Its major features include powerful full-text search, hit highlighting, faceted search, near real-time indexing, dynamic clustering, database integration, rich document (e.g., Word, PDF) handling, and geospatial search."],
    'codeserver'   :["Git Version Control","Git is a free and open source distributed version control system designed to handle everything from small to very large projects with speed and efficiency."]
};

App.Router.map(function() {
  // put your routes here
  //this.route("diagram", {path:"/diagram"});
  this.resource('diagram', { path: 'diagram/:env_id' });
});

App.IndexRoute = Ember.Route.extend({
  beforeModel: function() {
    this.transitionTo('diagram');
  }
});

App.DiagramRoute = Ember.Route.extend({
  model: function(params){
    return new Ember.RSVP.Promise(function(resolve){
      App.env = params.env_id;
      var target = install_dir + "/index.php"
      //local testing data or static data
      if(location.hostname == "localhost"){
        target = "data/servers.json"
      }
      if(App.env == 'enterprise'){
        target = install_dir + "data/enterprise.json";
      }
      else if(App.env == 'business'){
        target = install_dir + "data/business.json";
      }
      else if(App.env == 'personal'){
        target = install_dir + "data/personal.json";
      }

      $.get("/"+target+"?env="+params.env_id, function(d){
      if(!App.mySite){
          App.mySite = new Site();
        }
        App.mySite.load(d.servers);
        resolve(App.mySite);
        if(App.poller){
          clearInterval(App.poller);
        }
        App.poller = window.setInterval(function(){
          $.get("/"+target+"?env="+params.env_id, function(d){
            App.mySite.load(d.servers);
          }, "json");
        },2000);
      }, "json");
    });
  },
  setupController: function (controller, model) {
    controller.set('model', model);
    controller.set('graph', false);
  }
});

var infobarViewInstance, infobarContentViewInstance;

App.NavView = Ember.View.extend({
  templateName: 'nav',
  didInsertElement: function(){
    var self = this;
    $('li.link').removeClass('active');
    $('li.link.'+App.env).addClass('active');
    $(window).on('hashchange', function() {
      $('li.link').removeClass('active');
      $('li.link.'+App.env).addClass('active');
    });
  },
});

App.InfobarView = Ember.View.extend({
  templateName: 'infobarContainer',
  didInsertElement: function(){
    var self = this;
    infobarViewInstance = self;
  },
  open: function(){
    var self = this;
    //$("#infobarContainer").show("slide", {direction:"right"}, 500);
    $("#infobarContainer").addClass("emerge");
    $("#ember248").addClass("emerge");
  },
  close: function(){
    var self = this;
    //$("#infobarContainer").hide("slide", {direction:"right"}, 500);
    $("#infobarContainer").removeClass("emerge");
    $("#ember248").removeClass("emerge");
  }
});

App.InfobarContentView = Ember.View.extend({
  templateName: 'infobarContent',
  didInsertElement: function(){
    var self = this;
    infobarContentViewInstance = self;
    self.set("controller.title", "");
    self.set("controller.text", "");

    if(self.get("controller.node")){
      self.nodeChange();
    }
  },
  nodeChange : function(){
    var self = this;
    node = self.get("controller.node");
    if(node){
      self.set("controller.title", _marketing[node.type][0]);
      self.set("controller.text", _marketing[node.type][1]);
      var icoColor = "white";
      self.set("controller.icon", "/platform/img/" + node.type + "-white-72x72.png");
      $("#infobarContainer").removeClass().addClass(node.type);
      $(".infobarContent").find("img.headerIcon").attr("src", "/platform/img/" + node.type + "-" + icoColor + "-72x72.png");
      infobarViewInstance.open();
    }
    else{
      infobarViewInstance.close();
    }
  }.observes("controller.node")
});

App.DiagramView = Ember.View.extend({
  templateName: 'diagram',
  didInsertElement: function(){
    var self = this;
    var graph = self.get("controller.graph");
    var mySite = App.mySite;//self.get('controller.model');
    var elementId = "#diagramContainer";
    //Set dimension of graphcontainer.
    var graphWidth = Math.min($("body").width(), maxWidth) - $("#infobarContainer").width();
    var graphHeight = 550;//$("body").height();
    // if(graphHeight < 600){
    //   graphHeight = 600;
    // }
    $(diagramContainer).width(graphWidth).height(graphHeight);
    $("#infobarContainer").height($("body").height());

    if(!graph){
      mySite.registerEvent("server.add",function(e){
        self.drawServer(e.server);
        self.drawAllLinks();
      });

      mySite.registerEvent("server.delete",function(e){
        self.undrawServer(e.server);
      });

      graph = new myGraph(elementId);
      graph.registerEvent("node.selected", function(e){
        infobarContentViewInstance.set("controller.node", e.node);
      });
      graph.registerEvent("node.unselected", function(e){
        if(infobarViewInstance.get("controller.node") == e.node){
          infobarContentViewInstance.set("controller.node", false);
        }
      });
      self.set('controller.graph', graph);
      self.draw();
    }
  },
  draw: function(){
    var self = this;
    var graph = self.get("controller.graph");
    var mySite = App.mySite;//self.get('controller.model');
    var instances = mySite.get([]);
    $.each(instances, function(i,e){
      self.drawServer(e);
    });
    self.drawAllLinks();
  },
  undrawServer: function(server){
    var self = this;
    var graph = self.get("controller.graph");
    graph.removeNode(server.id);
  },
  drawServer: function(server){
    var self = this;
    var graph = self.get("controller.graph");
    graph.addNode(server);
  },
  drawAllLinks: function(){
    var self = this;
    var mySite = App.mySite;//self.get('controller.model');
    var instances = mySite.get([]);
    $.each(instances, function(i,e){
      self.drawLinks(e);
    });
  },
  drawLinks: function(server){
    var self = this;
    var graph = self.get("controller.graph");
    $.each(server.links, function(i,e){
      graph.addLink(server.id, e);
    });
  }
});


platform/js/graph.js
function myGraph(el) {
    var self = this;
    var classMap = {
        'edgeserver'   :"server",
        'appserver'    :"server",
        'dbserver'     :"server",
        'slavedbserver':"server",
        'cacheserver'  :"server",
        'fileserver'   :"server",
        'newrelic'     :"server",
        'indexserver'  :"server",
        'codeserver'   :"server"
    };

    var _types = ['edgeserver', 'appserver', 'dbserver', 'slavedbserver', 'cacheserver', 'fileserver', 'newrelic', 'indexserver', 'codeserver'];

    self._events = {
        'node.selected'  :[],
        'node.unselected':[]
    }

    this.registerEvent = function(event, cb){
        console.log("New Event Listener: ", event, cb);
        self._events[event].push(cb);
    }

    this._triggerEvent = function(event, target){
        var e = {
            "name": event,
            "node": target
        };
        console.log("Triggering Event: ", event, e);
        var listeners = self._events[event];
        $.each(listeners, function(i, cb){
            cb(e);
        });
    }

    // Add and remove elements on the graph object
    this.addNode = function (node) {
        console.log("adding node");
        if(node.type == "edgeserver"){
            node.fixed = true;
            node.x = w/2;
            node.y = 50;
        }
        else if(node.type == "newrelic"){
            node.fixed = true;
            node.x=300;
            node.y=150;
        }
        else if(node.type == "codeserver"){
            node.fixed = true;
            node.x=300;
            node.y=275;
        }
        nodes.push(node);
        update();
    }

    this.removeNode = function (id) {
        var i = 0;
        var n = findNode(id);
        while (i < links.length){
            if ((links[i]['source'] == n)||(links[i]['target'] == n)){
                links.splice(i,1);
            }
            else{
                i++;
            }
        }
        nodes.splice(findNodeIndex(id),1);
        update();
    }

    this.addLink = function (source, target){
        // if(source.type == "newrelic" || target.type == "newrelic"){
        //     // Don't factor these links in, they screw things up!
        //     return;
        // }
        if(findNode(source) && findNode(target)){
            links.push({"source":findNode(source),"target":findNode(target)});

            update();
        }
    }

    var findNode = function(id){
        var toReturn = false;
        $.each(nodes, function(i, obj){
            if(obj.id == id){
                toReturn = obj;
            }
        });
        return toReturn;
    }

    var findNodeIndex = function(id){
        for (var i in nodes) {if (nodes[i]["id"] === id) return i};
    }

    var bg = [
        {"id":"edgeserver", "pos":50},
        {"id":"appserver", "pos":200}
    ];

    var color = d3.scale.category20();

    // set up the D3 visualisation in the specified element
    var w = $(el).innerWidth(),
        h = $(el).innerHeight(),
        r = 36, //Circle radius
        st = 0,
        bgcolor = 0,
        sw = 0,
        fill = "#313945";
        //sc = "";

    var vis = this.vis = d3.select(el).append("svg:svg")
        .attr("width", w)
        .attr("height", h)
        .attr("stroke", st)
        .attr("background", bgcolor)
        .attr("fill", fill)
        .attr("stroke-width", sw);
        //.attr("class", sc);

    var bgGroup = vis.insert("g").attr("class", "bgGroup");
    var linkGroup = vis.insert("g").attr("class", "linkGroup");
    var nodeGroup = vis.insert("g").attr("class", "nodeGroup");

    // Remember the node that is selected.
    var selectedNode = false;

    var force = d3.layout.force()
        //.chargeDistance(.10000)
        //.alpha(0)
        .gravity(1)
        .friction(.7)
        .linkStrength(function(d){
            if(d.target.type == "newrelic" || d.target.type == "codeserver"){
                return 0;
            }
            return 1;
        })
        .distance(function(d){
            //var dist = 10 + (10*d.source.links.length)
            //return dist;
            if(d.target.type == "codeserver" || d.target.type == "newrelic" || d.source.type == "codeserver" || d.source.type == "newrelic"){
                console.log("d.target", d.target);
                return w/2;
            }
            if(d.source.links.length > 5 || d.target.type == "appserver"){
                return 150;
            }
            return 150;
        })//Could be a function
        .charge(function(d){
            if(d.type == "appserver"){
                return -20000;
            }
            else if(d.type=="newrelic" || d.type=="codeserver"){
                return 0;
            }
            return -20000;
        })//Could be a function
        .size([w, h]);

    var nodes = force.nodes(),
        links = force.links();

    vis.on("click", function(){
        if(d3.event.defaultPrevented) return;
        console.log("bg click");
        if(selectedNode){
            self._triggerEvent("node.unselected",selectedNode);
            selectedNode = false;
            activate();
        }
        d3.event.stopPropagation();
    });
    var activate;
    var update = function(){

        //Draw the bg
        var bgLine = bgGroup.selectAll("g.bg")
            .data(bg, function(d){return d.id;});

        bgLine.enter().insert("g")
            .attr("class", "bg link")
            .attr("x1", function(d) { return 0; })
            .attr("x2", function(d) { return w; })
            .attr("y1", function(d) { return d.pos; })
            .attr("y2", function(d) { return d.pos; });

        bgLine.exit().remove();

        //Draw the nodes
        var node = nodeGroup.selectAll("g.node")
            .data(nodes, function(d) { return d.id;})
            .on("click", function(d) {
                if(d3.event.defaultPrevented) return; // ignore drag
                console.log("node clicked!");
                nodeGroup.selectAll("g.node").attr("stroke", "#EFD01B");
                nodeGroup.selectAll("g.node").attr("background", "#313945");
                nodeGroup.selectAll("g.node").attr("fill", "#313945");
                nodeGroup.selectAll("g.node").attr("stroke-width", 0);
                //nodeGroup.selectAll("g.node").attr("class", "");
                if(selectedNode.id == d.id){
                    //Unselect if selected node is already selected, duh!
                    console.log("Unselecting");
                    selectedNode = false;
                    self._triggerEvent("node.unselected",d);
                }
                else{
                    console.log("Selecting");
                    selectedNode = d;
                    d3.select(this).attr("stroke", "#EFD01B");
                    d3.select(this).attr("stroke-width", 4);
                    d3.select(this).attr("background", "#00A9E0");
                    d3.select(this).attr("fill", "#00A9E0");
                    //d3.select(this).attr("class", "svgClass");
                    self._triggerEvent("node.selected", d);
                }

                activate();//force.resume();//update();
                d3.event.stopPropagation();
            });

        var nodeEnter = node.enter().insert("g")
            .attr("class", "node")
            //.call(force.drag);

        nodeEnter.append("circle")
            .attr("class", "bgCircle")
            .attr("r", r)

        nodeEnter.append("circle")
            .attr("class", function(d){ return "server "+d.type; })
            .attr("r", r)

        nodeEnter.append("image")
            .attr("xlink:href", function(d){ return "img/" + d.type + "-white-72x72.png"; })
            .attr("class", "icon")
            .attr("x", -22).attr("y", -22)
            .attr("width", 44).attr("height", 44)

        node.exit().remove();


        //Draw the links
        var link = linkGroup.selectAll("line.link")
            .data(links, function(d) { return d.source.id + "-" + d.target.id; });

        link.enter().insert("line")
            .attr("class", function(l){
                classes = ["link"];
                if(l.target.type == "newrelic" || l.target.type=="codeserver"){
                    classes.push("dashed");
                }
                return classes.join(" ");
            });

        link.exit().remove();


        //Activate Links and Nodes based on node selection - called be event handlerss
        activate = function(){
            //Highlight pertinent links
            //Sorry about the verbosity of this... I guess I am too tired to get this right without typing it out.
            var targetsOrSources = [];
            link.classed("active",function(d){
                var active = false;
                if(d.target.id == selectedNode.id){active = true; }
                if(d.source.id == selectedNode.id){active = true; }
                if(active){
                    if(d.target.id != selectedNode.id){targetsOrSources.push(d.target); }
                    if(d.source.id != selectedNode.id){targetsOrSources.push(d.source); }
                }
                return active;
            });
            node.classed("inactive",function(n){
                //No selection, bail.
                if(!selectedNode){return false;}
                if(targetsOrSources.indexOf(n) > -1){return false; }
                if(n.id == selectedNode.id){return false; }
                return true;
            });
        }

        //Updating on ticks
        force.on("tick", function(e) {
            // var desiredWidth = 1200;
            // var maxX = desiredWidth/2 + desiredWidth/4;
            // var minX = desiredWidth/2 - desiredWidth/4;

            node.attr("transform", function(d) { return "translate(" + d.x + "," + d.y + ")"; });

            // node.attr("cx", function(d) { return d.x = Math.max(minX+r, Math.min(maxX - r, d.x)); })
            //     .attr("cy", function(d) { return d.y = Math.max(r, Math.min(h - r, d.y)); });
            node.attr("cx", function(d) { return d.x = Math.max(r, Math.min(w - r, d.x)); })
                .attr("cy", function(d) { return d.y = Math.max(r, Math.min(h - r, d.y)); });

            nodes.forEach(function(d, i){
                //if(d.type == "edgeserver"){d.y = 50; d.x = w/2;}
                /*else*/ if(d.type == "appserver"){d.y = 200;}
                else if(d.type == "dbserver" || d.type == "fileserver" || d.type == "cacheserver" || d.type == "indexserver"){d.y = 350;}
                else if(d.type == "slavedbserver"){d.y = 500;}
                //else if(d.type == "codeserver"){d.x=w-200; d.y=275}
                //else if(d.type == "newrelic"){d.x=w-100; d.y=50}
                //else{d.y = 500;}
            });

              // var q = d3.geom.quadtree(nodes),
              //     i = 0,
              //     n = nodes.length;

              // while (++i < n) {
              //   q.visit(collide(nodes[i]));
              // }

            var k = 20 * e.alpha;
            // links.forEach(function(d, i) {
            //     //Push up the source of the link a little.
            //     //If it is the source of many it will be forced to the top of the vis.
            //     //d.source.y -= k;
            //     //d.target.y += k;

            //     // //Make sure the edge servers float to the top.
            //     // if(d.source.type == "edgeserver"){
            //     //     d.source.y -= (1*k);
            //     // }

            // });

            link.attr("x1", function(d) { return d.source.x; })
                .attr("y1", function(d) { return d.source.y; })
                .attr("x2", function(d) { return d.target.x; })
                .attr("y2", function(d) { return d.target.y; });
        });

        var collide = function(node){
          var r = node.radius + 30,
              nx1 = node.x - r,
              nx2 = node.x + r,
              ny1 = node.y - r,
              ny2 = node.y + r;
          return function(quad, x1, y1, x2, y2) {
            if (quad.point && (quad.point !== node)) {
              var x = node.x - quad.point.x,
                  y = node.y - quad.point.y,
                  l = Math.sqrt(x * x + y * y),
                  r = node.radius + quad.point.radius;
              if (l < r) {
                l = (l - r) / l * .5;
                node.x -= x *= l;
                node.y -= y *= l;
                quad.point.x += x;
                quad.point.y += y;
              }
            }
            return x1 > nx2
                || x2 < nx1
                || y1 > ny2
                || y2 < ny1;
          };
        }

        // Restart the force layout.
        force.start();
    }

    // Make it all go
    update();
}
